triggers/service/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Response.scala

```diff
@@ -22,32 +23,38 @@ object Response {
 
   // These functions are borrowed from the HTTP JSON ledger API but I haven't
   // factored them out for now as they are fairly small.
-  def errorsJsObject(status: StatusCode, es: Seq[String]): JsObject = {
+  private[this] def errorsJsObject(status: StatusCode, es: Seq[String]): JsObject = {
     val errors = es.toJson
     JsObject(statusField(status), ("errors", errors))
   }
 
-  def resultJsObject[A: JsonWriter](a: A): JsObject = {
+  private[this] def resultJsObject[A: JsonWriter](a: A): JsObject = {
     resultJsObject(a.toJson)
   }
 
-  def resultJsObject(a: JsValue): JsObject = {
+  private[this] def resultJsObject(a: JsValue): JsObject = {
     JsObject(statusField(StatusCodes.OK), ("result", a))
   }
 
-  def statusField(status: StatusCode): (String, JsNumber) =
+  private[this] def statusField(status: StatusCode): (String, JsNumber) =
     ("status", JsNumber(status.intValue()))
 
   // Trigger status messages have timestamps for which this is the
   // formatter.
-  object LocalDateTimeJsonFormat extends RootJsonFormat[LocalDateTime] {
-    override def write(dt: LocalDateTime) =
-      JsString(dt.format(DateTimeFormatter.ISO_LOCAL_DATE_TIME))
-
-    override def read(json: JsValue): LocalDateTime = json match {
-      case JsString(s) => LocalDateTime.parse(s, DateTimeFormatter.ISO_LOCAL_DATE_TIME)
-      case _ => throw new DeserializationException("Decode local datetime failed")
+  implicit val LocalDateTimeJsonFormat: RootJsonFormat[LocalDateTime] =
```

Comment:
For a library, there are interesting good reasons related to API stability and proper separation of interface and implementation.  I won't go into those, but focus on a semantics difference affecting all use cases:

```scala
Welcome to Scala 2.12.10 (OpenJDK 64-Bit Server VM, Java 1.8.0_202).
Type in expressions for evaluation. Or try :help.

scala> class Foo[T]
defined class Foo

scala> object Qqq {
     |   implicit val a: Foo[Int] = new Foo
     |   implicit def b: Foo[Int] = new Foo
     | }
defined object Qqq

scala> import Qqq._
import Qqq._

scala> implicitly[Foo[Int]]
<console>:17: error: ambiguous implicit values:
 both value a in object Qqq of type => Foo[Int]
 and method b in object Qqq of type => Foo[Int]
 match expected type Foo[Int]
       implicitly[Foo[Int]]
                 ^
```

This error is desirable; you have multiple instances in scope and Scala refuses to pick one.

Consider the opposite implementation, though:

```scala
scala> object Qqq {
     |   implicit object a extends Foo[Int]
     |   implicit def b: Foo[Int] = new Foo
     | }
defined object Qqq

scala> import Qqq._
import Qqq._

scala> implicitly[Foo[Int]]
res1: Foo[Int] = Qqq$a$@25031fd4
```

Even though we have conflicting instances, scalac will pick one.  Why?  There's a hint here:

```scala
scala> implicitly[a.type]
res3: Qqq.a.type = Qqq$a$@25031fd4
```

but if we go back to the original code:

```scala
scala> implicitly[a.type]
<console>:24: error: could not find implicit value for parameter e: Qqq.a.type
       implicitly[a.type]
                 ^

scala> a: a.type
res5: Qqq.a.type = Foo@5bee88e3
```

An `implicit object bar extends Baz` is not *just* an implicit `Baz`, it is also an implicit `bar.type`, a strict subtype of `Baz`.  In the above, `a.type` exists regardless, but it is only implicit *itself* if you use `implicit object`.  And that causes SLS [§7.2](https://scala-lang.org/files/archive/spec/2.12/07-implicits.html#implicit-parameters), [§6.26.3](https://scala-lang.org/files/archive/spec/2.12/06-expressions.html#overloading-resolution) to kick in.

This can be harmless.  *Or*, you've imported `JsonFormat`s from multiple places and suddenly one of them silently works instead of giving you a conflict error.

Reply:
Makes sense, thanks for writing this up!